### PR TITLE
Update -  Using diff on register and update logic

### PR DESCRIPTION
**Pull request by at-haingo**
#### Related issues (Required)
- Task #242  #254 
- Bug #

#### Description (Required)
- [x] Using diff on register 
- [x] Update logic

### CODE REVIEW

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt [hunk 1]
```diff
@@ -3,33 +3,54 @@ package vn.asiantech.way.ui.register
...
     internal val backStatus: PublishSubject<Boolean> = PublishSubject.create()
+    internal val countryObservable: SingleSubject<Country> = SingleSubject.create()
+    internal val countries = mutableListOf<Country>()
     private var lastClickTime = 0L
     private lateinit var user: User
 
     constructor(context: Context, isRegister: Boolean) : this(WayRepository(), LocalRepository(context), isRegister)
 
-    internal fun getCountries(): Observable<List<Country>> = assetDataRepository.getCountries()
+    internal fun getCountries(): Observable<DiffUtil.DiffResult> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/a843bceb60b5154a7d86539c28978a9fdb9fe573/app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt#L37"></span></td>
    </tr>
  </tbody>
</table>

**Comment by at-dev-support-01** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please rebase to get rid of the merge commits in this PR</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">All test passed. (21/21)</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt [hunk 2]
```diff
@@ -3,33 +3,54 @@ package vn.asiantech.way.ui.register
...
     internal val backStatus: PublishSubject<Boolean> = PublishSubject.create()
+    internal val countryObservable: SingleSubject<Country> = SingleSubject.create()
+    internal val countries = mutableListOf<Country>()
     private var lastClickTime = 0L
     private lateinit var user: User
 
     constructor(context: Context, isRegister: Boolean) : this(WayRepository(), LocalRepository(context), isRegister)
 
-    internal fun getCountries(): Observable<List<Country>> = assetDataRepository.getCountries()
+    internal fun getCountries(oldCountries: List<Country> = countries): Observable<DiffUtil.DiffResult> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/193a71546c42f4cc6f50e5da34f909639fc6d267/app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt#L37"></span></td>
    </tr>
  </tbody>
</table>

File: app/src/test/java/vn/asiantech/way/register/RegisterViewModelTest.kt [hunk 3]
```diff
@@ -83,21 +86,120 @@ class RegisterViewModelTest {
     }
 
     @Test
-    fun `Given countries - When get countries - Then return right countries`() {
+    fun `Given null oldCountries - When get countries - Then return right diff`() {
```
Comment (at-vinhhuynh) [inline app/src/test/java/vn/asiantech/way/register/RegisterViewModelTest.kt]: `oldCountries` change to `old countries`

File: app/src/test/java/vn/asiantech/way/register/RegisterViewModelTest.kt [hunk 4]
```diff
@@ -83,21 +86,120 @@ class RegisterViewModelTest {
...
+        Assert.assertThat(viewModel.countries[0].iso, `is`("84"))
+        Assert.assertThat(viewModel.countries[0].tel, `is`("12345"))
+        Assert.assertThat(viewModel.countries[1].iso, `is`("12"))
+        Assert.assertThat(viewModel.countries[1].tel, `is`("123456789"))
+    }
+
+    @Test
+    fun `Given oldCountries already value - When get countries - Then return right diff`() {
+        /* Given */
+        val oldCountries = listOf(Country("84", "12345"))
```
Comment (at-vinhhuynh) [inline app/src/test/java/vn/asiantech/way/register/RegisterViewModelTest.kt]: remove this variable

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt (hunk 1 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/90eac17cfd2541a41b043643a7f51d655f402eae/app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt#L37"></span></td>
    </tr>
  </tbody>
</table>

**Comment by at-vinhhuynh** [comment]
@at-tienhoang 
I finish review this PR, please check it again.
thank you!

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt [hunk 5]
```diff
@@ -51,6 +72,16 @@ class RegisterViewModel(private val wayRepository: WayRepository, private val as
         return wayRepository.getUser()
                 .doOnSuccess {
                     user = it
+                    val basePhone: List<String>? = user.phone?.split("/")
+                    if (basePhone != null) {
+                        val isoCode = basePhone[0]
+                        for (i in 0 until countries.size) {
```
Comment (at-tienhoang) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt]: create ext function for this
MutableList<Country>.getCountryByIso

File: app/src/main/java/vn/asiantech/way/extension/MutableList.kt [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package vn.asiantech.way.extension
+
+import vn.asiantech.way.data.model.Country
+
+/**
+ *
+ * Created by haingoq on 22/12/2017.
+ */
+fun MutableList<Country>.getCountryByIso(iso: String): Country? {
+    val country = filter { iso == it.iso }[0]
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/extension/MutableList.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/d2f476d53ba765ab91e8f3ac0475f56d26a1c7d0/app/src/main/java/vn/asiantech/way/extension/MutableList.kt#L10"></span></td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt [hunk 7]
```diff
@@ -3,33 +3,55 @@ package vn.asiantech.way.ui.register
...
     internal val backStatus: PublishSubject<Boolean> = PublishSubject.create()
+    internal val countryObservable: SingleSubject<Country> = SingleSubject.create()
+    internal val countries = mutableListOf<Country>()
     private var lastClickTime = 0L
     private lateinit var user: User
 
     constructor(context: Context, isRegister: Boolean) : this(WayRepository(), LocalRepository(context), isRegister)
 
-    internal fun getCountries(): Observable<List<Country>> = assetDataRepository.getCountries()
+    internal fun getCountries(): Observable<DiffUtil.DiffResult> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/d2f476d53ba765ab91e8f3ac0475f56d26a1c7d0/app/src/main/java/vn/asiantech/way/ui/register/RegisterViewModel.kt#L38"></span></td>
    </tr>
  </tbody>
</table>

**Pull request merged**